internal/config/app_config_sort.go

```diff
@@ -0,0 +1,182 @@
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// Simple wrapper around the hclwrite call to turn a traversal into a string.
+func renderTraversal(t hcl.Traversal) (string, error) {
+	tokens := hclwrite.TokensForTraversal(t)
+	var buf bytes.Buffer
+
+	_, err := tokens.WriteTo(&buf)
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// Captures the information about a variable calculated during sort to be used
+// without recalculation by the caller.
+type analyzedPair struct {
+	Pair     hcl.KeyValuePair
+	Name     string
+	Internal bool
+	Refs     []string
+}
+
+type VariableLoopError struct {
+	LoopVars []string
+}
+
+func (v *VariableLoopError) Error() string {
+	return fmt.Sprintf("loop detected amongst variables: %s", strings.Join(v.LoopVars, ", "))
+}
+
+// sortVars performs a topological sort on EnvRaw and InternalRaw. The sort
+// yields the pairs in most referenced to least referenced order. Meaning
+// that the if pair X references pair R, then R will be before X in the slice.
+func (c *genericConfig) sortVars(ctx *hcl.EvalContext) ([]*analyzedPair, error) {
```

Comment:
I think this needs unit tests. Given its a pure function that is executing a well known algorithm, I think we can cover all the edge cases and ensure this really works. I know its tested incidentally through higher level tests, but they're mostly happy path cases.

Reply:
Yup, added a few. There aren't a ton of edge cases so it should cover them.